pkg/controllers/deprovisioning/emptiness.go

```diff
@@ -0,0 +1,109 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deprovisioning
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/utils/clock"
+
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/samber/lo"
+
+	"github.com/aws/karpenter-core/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+	"github.com/aws/karpenter-core/pkg/metrics"
+)
+
+// Emptiness is a subreconciler that deletes empty nodes.
+// Emptiness will respect TTLSecondsAfterEmpty
+type Emptiness struct {
+	kubeClient client.Client
+	clock      clock.Clock
+	cluster    *state.Cluster
+}
+
+// shouldNotBeDeprovisioned is a predicate used to filter deprovisionable nodes
+func (e *Emptiness) shouldNotBeDeprovisioned(ctx context.Context, n *state.Node, provisioner *v1alpha5.Provisioner, nodePods []*v1.Pod) bool {
+	if provisioner == nil || provisioner.Spec.TTLSecondsAfterEmpty == nil {
+		return true
+	}
+
+	emptinessTimestamp, hasEmptinessTimestamp := n.Node.Annotations[v1alpha5.EmptinessTimestampAnnotationKey]
+	ttl := time.Duration(ptr.Int64Value(provisioner.Spec.TTLSecondsAfterEmpty)) * time.Second
+
+	emptinessTime, err := time.Parse(time.RFC3339, emptinessTimestamp)
+	if err != nil {
+		logging.FromContext(ctx).Debugf("Unable to parse emptiness timestamp, %s for node %s", emptinessTimestamp, n.Node.Name)
+		return true
+	}
+	// Don't deprovision if node is not empty, does not have the emptiness timestamp, or is before the emptiness TTL
+	return len(nodePods) != 0 || !hasEmptinessTimestamp || !e.clock.Now().After(emptinessTime.Add(ttl))
+}
+
+// sortCandidates orders deprovisionable nodes by the disruptionCost
+func (e *Emptiness) sortCandidates(nodes []candidateNode) []candidateNode {
+	return nodes
+}
+
+// computeCommand generates a deprovisioning command given deprovisionable nodes
+func (e *Emptiness) computeCommand(_ context.Context, _ int, nodes ...candidateNode) (deprovisioningCommand, error) {
+	emptyNodes := lo.Filter(nodes, func(n candidateNode, _ int) bool { return len(n.pods) == 0 })
+	if len(emptyNodes) == 0 {
+		return deprovisioningCommand{action: actionDoNothing}, nil
+	}
+	return deprovisioningCommand{
+		nodesToRemove: lo.Map(emptyNodes, func(n candidateNode, _ int) *v1.Node { return n.Node }),
```

Comment:
This won't be able to execute now, right? Since we fail the execution if there is more than one node in this list at present?